File: .algorithm_and_structure/isqrt.cpp
// ビット管理
ll isqrt(ll n) {
	ll i = 1;
	while (i * i < n) {
		i <<= 1;
	}
	ll res = 0;
	while (i > 0) {
		if ((res + i) * (res + i) <= n) {
			res += i;
		}
		i >>= 1;
	}
	return res;
}

File: .algorithm_and_structure/isqrt.py
def isqrt(num):
	assert(0 <= num)
	left = -1
	right = num+1
	while right - left > 1:
		mid = (left + right) // 2
		if mid ** 2 <= num:
			left = mid
		else:
			right = mid
	return left
